### ISSUE TITLE
Weather data 

### ISSUE THREAD

**Issue by mvisani** [earth-metabolome-initiative/emi-monorepo#29]
We need to find a way that given a geolocation, we retrieve: 
- Rain
- Cloudiness 
- Temperature
- Altitude ?

**Comment by LucaCappelletti94** [comment]
Hi Marco, start by:

* [ ] Searching for crates that provide metereological data
* [ ] Maybe there is already a postgres extension for this type of datasets? If there is one, we can chat to run the build.rs of the codegen on it and see which functions it makes available.
* [ ] Define a table that captures the complexity of those datasets, potentially multiple tables.
* [ ] Create a crate to model this and start to run it on a few reasonable test cases (e.g. wheather a week ago)
* [ ] Identify the limits of time and resolution of this system. Search on google extensively for corner cases in the APIs you choose to use. It is very likely you will need to query different services for different geographical areas. Model the code in such a way to model different services avoiding excessive code redoundancy.

Let me know by end of day what is your impression on the time requirements to get these points done, please.

Luca

**Comment by mvisani** [comment]
Adding some example links here : 
- https://open-meteo.com/en/docs/historical-weather-api

**Comment by mvisani** [comment]
@LucaCappelletti94 I found [this crate](https://crates.io/crates/open-meteo-rs) that works nicely. Should we discuss how we want to integrate the data with ours ? Is the crate sufficient ? Is it a good starting point for creating our own crate ?

**Comment by LucaCappelletti94** [comment]
@mvisani I took a look at the crate and it has a very small amount of downloads, but at least the repo has a few stars. It only relies on a single weather service, so we would need to understand whether that weather service has a price, which part of the world it covers and so on.

For the time being, write a binary that tests a random location for each continent, and see which issues this may illustrate (maybe they do not have data for some places, depending on their source of data):

- Western Sahara
- Greenland
- Some random area in the Pacific
- China

Also check how far back these data goes.

**Comment by mvisani** [comment]
Regarding Open Meteo [here](https://open-meteo.com/en/docs/historical-weather-api) is what they say about their historical weather API. Data goes back to 1940, and all regions of the world are included. 
The terms of use of the website for non-commercial use are : 

> - Less than 10'000 API calls per day, 5'000 per hour and 600 per minute.
> - You may only use the free API services for non-commercial purposes.
> - You accept to the CC-BY 4.0 license, as specified in the [license conditions](https://open-meteo.com/en/license).
> - We reserve the right to block applications and IP addresses that misuse our service without prior notice. 

![Image](https://github.com/user-attachments/assets/909b3cfa-68f6-43d6-99ab-0b9a66017a7b)

**Comment by LucaCappelletti94** [comment]
Have you looked into the self-hosted version? Can you setup a rust binary to provide some documented examples?

**Comment by LucaCappelletti94** [comment]
Stuff to do as per call:

* [ ] create a crate in the monorepo under `/web` in a new directory called `/api`
* [ ] in there prepare the code necessary to call to the weather data (using any available crates that may be needed)
* [ ] after you define a test suite to verify that everything works fine, define the SQL schema to structure the weather data
* [ ] once the above is done, using [`pg_cron`](https://github.com/citusdata/pg_cron) and [`pgrx`](https://github.com/pgcentralfoundation/pgrx), setup a mechanism that once a new row is inserted in the expected tables with locations, an API call is executed and the result is stored.

**Comment by mvisani** [comment]
Regarding the SQL schema I think it should look like this : 
```sql
CREATE TABLE sample_weather_data (
    weather_id PRIMARY KEY,
    sample_id INT NOT NULL,
    requested_at TIMESTAMP,  -- When we fetched the data
    temperature_2m FLOAT,
    ...,
    FOREIGN KEY (sample_id) REFERENCES samples(sample_id)
```

One thing that is still not clear to me is how do we handle the different units (speed, temperature, etc). Should it be a reference to a different table ? What about the conversions between say km/h or m/s ?

**Comment by LucaCappelletti94** [comment]
There is a `units` table, but here units should not be relevant as you should be inserting information strictly following [`ISU`](https://en.wikipedia.org/wiki/International_System_of_Units) - that is, `m/s` for instance. Please fully fill in the `...` you left out, the weather data themselves are still left unspecified. I am not sure whether we need a `weather_id` primary key, as the `sample_id` may already function effectively as a primary key.

Also, it may be the case that several samples collected all roughly at the same time and location may all share the same weather data entry - what is the time and location granularity of the information of these API?

**Comment by mvisani** [comment]
I've updated the table (see above).

The API has a granularity up to the hour. However the API only accepts requests up to the day. So when requesting hourly data, the API will return an array of 24 data points. We could then get the closest hourly datapoint of that array.

The smallest spatial resolution is 9km and highest is . So samples will definitely share weather data entries.

**Comment by LucaCappelletti94** [comment]
If sample collections' associated weather is shared across all collections within a given geographical area and time, it may be better to have one such entry in an appropriate table (a table with spacetime coordinates basically) and then have this weather table link to that, maybe as [an inheritance](https://www.postgresql.org/docs/current/ddl-inherit.html). 

Right now all of the columns you have added are optional - is this correct? Try to retrieve weather data for a few locations and discover which entries are actually commonly not provided, it may help prune away some of the columns that are actually consistently not provided.

Also, the table right now is very large (has more than 32 columns) - we should try and avoid making tables with more than 16 columns if at all possible.

**Comment by mvisani** [comment]
> Right now all of the columns you have added are optional - is this correct? Try to retrieve weather data for a few locations and discover which entries are actually commonly not provided, it may help prune away some of the columns that are actually consistently not provided.

Yes they are all optional but they all return values if they are selected. I would then need to add `NOT NULL` to the columns right ? If we need to reduce the number of columns, we can either discuss if some of those values are really necessary for our use case, or we could split the data into multiple tables ?
